### ISSUE TITLE
all(...) rewrite produces incorrect results

### ISSUE THREAD

**Issue by asottile** [pytest-dev/pytest#5372]
this is boiled down from a more complicated case

```python
def test():
    y = range(5)
    assert not all(x for x in y)
```

```console
$ pytest t.py
============================= test session starts ==============================
platform linux -- Python 3.6.7, pytest-4.6.1, py-1.8.0, pluggy-0.12.0
rootdir: /tmp
collected 1 item                                                               

t.py F                                                                   [100%]

=================================== FAILURES ===================================
_____________________________________ test _____________________________________

    def test():
        y = range(5)
>       assert not all(x for x in y)
E       assert 0

t.py:3: AssertionError
=========================== 1 failed in 0.04 seconds ===========================
$ pytest t.py --assert=plain
============================= test session starts ==============================
platform linux -- Python 3.6.7, pytest-4.6.1, py-1.8.0, pluggy-0.12.0
rootdir: /tmp
collected 1 item                                                               

t.py .                                                                   [100%]

=========================== 1 passed in 0.01 seconds ===========================```

**Comment by tacaswell** [comment]
We also ran into issues with list comprehensions inside of all:

```python

def test_foobar():
    events = [{'data': d} for d in [{'modified_motor': -5.0, 'modified_motor_setpoint': -5.0, 'modified_det': -3.726653172078671e-06}, {'modified_motor': -4.0, 'modified_motor_setpoint': -4.0, 'modified_det': -0.00033546262790251185}, {'modified_motor': -3.0, 'modified_motor_setpoint': -3.0, 'modified_det': -0.011108996538242306}, {'modified_motor': -2.0, 'modified_motor_setpoint': -2.0, 'modified_det': -0.1353352832366127}, {'modified_motor': -1.0, 'modified_motor_setpoint': -1.0, 'modified_det': -0.6065306597126334}, {'modified_motor': -0.0, 'modified_motor_setpoint': -0.0, 'modified_det': -1.0}, {'modified_motor': -1.0, 'modified_motor_setpoint': -1.0, 'modified_det': -0.6065306597126334}, {'modified_motor': -2.0, 'modified_motor_setpoint': -2.0, 'modified_det': -0.1353352832366127}, {'modified_motor': -3.0, 'modified_motor_setpoint': -3.0, 'modified_det': -0.011108996538242306}, {'modified_motor': -4.0, 'modified_motor_setpoint': -4.0, 'modified_det': -0.00033546262790251185}]]
    #     dis.dis(test_foobar)

    v = all([evt['data'][key] <= 0
             for evt in events
             for key in evt['data'].keys()])
    assert v
    assert all([evt['data'][key] <= 0
                for evt in events
                for key in evt['data'].keys()])
    key = 1
    assert key

```

```
=========================================================== test session starts ============================================================
platform linux -- Python 3.7.3, pytest-4.6.1, py-1.8.0, pluggy-0.12.0
rootdir: /tmp
plugins: cov-2.6.0, faulthandler-1.5.0, rerunfailures-5.0, timeout-1.3.2, forked-0.2, xdist-1.24.0
collected 1 item                                                                                                                           

test2.py F                                                                                                                           [100%]

================================================================= FAILURES =================================================================
_______________________________________________________________ test_foobar ________________________________________________________________

    def test_foobar():
        events = [{'data': d} for d in [{'modified_motor': -5.0, 'modified_motor_setpoint': -5.0, 'modified_det': -3.726653172078671e-06}, {'modified_motor': -4.0, 'modified_motor_setpoint': -4.0, 'modified_det': -0.00033546262790251185}, {'modified_motor': -3.0, 'modified_motor_setpoint': -3.0, 'modified_det': -0.011108996538242306}, {'modified_motor': -2.0, 'modified_motor_setpoint': -2.0, 'modified_det': -0.1353352832366127}, {'modified_motor': -1.0, 'modified_motor_setpoint': -1.0, 'modified_det': -0.6065306597126334}, {'modified_motor': -0.0, 'modified_motor_setpoint': -0.0, 'modified_det': -1.0}, {'modified_motor': -1.0, 'modified_motor_setpoint': -1.0, 'modified_det': -0.6065306597126334}, {'modified_motor': -2.0, 'modified_motor_setpoint': -2.0, 'modified_det': -0.1353352832366127}, {'modified_motor': -3.0, 'modified_motor_setpoint': -3.0, 'modified_det': -0.011108996538242306}, {'modified_motor': -4.0, 'modified_motor_setpoint': -4.0, 'modified_det': -0.00033546262790251185}]]
        #     dis.dis(test_foobar)
    
        v = all([evt['data'][key] <= 0
                 for evt in events
                 for key in evt['data'].keys()])
        assert v
>       assert all([evt['data'][key] <= 0
                    for evt in events
                    for key in evt['data'].keys()])
E       UnboundLocalError: local variable 'key' referenced before assignment

test2.py:10: UnboundLocalError
========================================================= 1 failed in 0.03 seconds =========================================================
(dd37) ✘  /tmp

```

attn @klauer who distilled this down to a MVP

**Comment by tacaswell** [comment]
Ah, I see #5370 already hit this case, sorry for the noise.

**Comment by asottile** [comment]
that one looks more like #5370 fwiw

**Comment by asottile** [comment]
this has been released as part of 3.6.2 -- those that were experiencing these issues may need to clean .pyc files after upgrading
